steps/24_categorical_support/step.md

```diff
@@ -0,0 +1,143 @@
+- The reason I proposed this is in case the user wants to treat a column with numerical values as categorical. This will not be detected by feature_kind as it relies only on the dtype.
+- More on how this is tackled in each design below.
+---
+---
+
+
+### Design 1. Allow categorical features ONLY in those estimators which support natively. (no encoding internally)
+
+- `categorical_features` parameter will be present only in estimators that support natively.
+- For estimators that do not support natively, users can make a pipeline with an encoder of their choice and use categorical variables. This is the general ml workflow, like how it's done in sklearn too.
+    
+    #### How tag is going to be used
+    - in datatype checks :
+    ```
+    if categorical support tag is false:
+        if categorical feature is present:
+            error - cat feature is not supported by this estimator
+    ```       
+
+- In this case `categorical feature` will not be a parameter in models that do not support. So (NO,YES,_) is not possible.    
+
+| **Estimator supports categorical natively** | **Cat feature specified by user** | **Cat feature present in data** | **Outcome** | 
+|:-----------------:|:-------------------------:|:--------------:|:---------------:|
+| NO | NO | NO | normal case, proceeds as usual without error. |
+| NO | NO | YES| informative error message to user that categorical feature is not supported by the estimator |
+| NO | YES| NO | X |
+| NO | YES| YES| X |
+| YES| NO | NO | normal case, proceeds as usual without error. |
+| YES| NO | YES| use categorical columns as detected by feature_kind and proceed. |
+| YES| YES| NO | informative error message that specified column was not present in data.|
+| YES| YES| YES| use categorical columns as detected by feature_kind along with the user specified columns(union) and proceed. |
+
+- in case where specified column was not found, we can also choose to ignore and continue with warning that column was not found.
+
+#### estimators that can be added:
+- xgboost, lightgbm and catboost are some models that have native support for categorical. But method of passing categorical features varies quite a lot between them (I've commented on this in umbrella issue - https://github.com/sktime/sktime/issues/6109). Making a general interface might be challenging.
+- So it might be better to interface them in separate classes inheriting from reduction forecaster class. (similar to darts lightgbm, see - https://unit8co.github.io/darts/generated_api/darts.models.forecasting.lgbm.html)
+
+#### PROS:
+- No internal encoding, comparatively easier to implement.
+- No major changes to core logic.
+- Follows general ML workflow i.e if estimator does not support, then user must preprocess their data(including encoding) and then use it.
+#### CONS:
+- Less number of estimators support categorical natively.
+
+
+---
+---
+
+### Design 2. Perform one hot encoding if not natively supported. 
```

Comment:
Notes: definitely can be default choice but should be configurable per column and/or overall.

Reply:
Is this really required though? Encoders exist for this very purpose and are available to users. Why do we have to provide all that functionality inbuilt? In sklearn, there is nothing of this sort. Users can just preprocess the data and then use it in estimators.

Reply:
In my opinion, it is really required.

- one-hot encoder is not applibale or suitable always.
  - variables with a lot of levels may create too many features, which can cause issues.
  - ordinal variables will loose complete order.
- if encoded outside, there will not be any possibility to use native support from the estimator when available in a pipeline step, and use encoding otherwise etc. I asked the pipeline question very much for this reason.
- since one hot encoder change the dimension of the data, it can actually be not supported in some cases or be complicated due to the collinearity it usually adds (unless you drop one column).

I don't think we should not do a direct comparison with `sklearn` in planning everything about `sktime`. They are not the same package, and the capabilities they provide are also not at same level mostly (`sklearn` implements all their algorithms, in most places `sktime` framework gives unified format from different underlying packages). This is very likely opinionated, and @fkiray and other core developers may not have the same view as well. But giving user the choice should be preferred over fixing something in my opinion.

Reply:
Yes of course. I agree that providing all options available is better than a single fixed encoding option. 

I meant, is the internal encoding necessary . And why?

Reply:
>  But giving user the choice should be preferred over fixing something in my opinion.

Precisely, and they have full choice on what columns to encode, what encoder to use and they can do it externally. I don't quite get why we want to make such major changes to sktime only to provide what can already be done.

Reply:
Indeed it's a complex change, there is no doubt about that at all. But that can't be the reason to not even consider in planning design phase, complexity of the task is why it was considered for full time summer project instead of open source contributors adding parts of it based on their availability.

> if encoded outside, there will not be any possibility to use native support from the estimator when available in a pipeline step, and use encoding otherwise etc. I asked the pipeline question very much for this reason.

I think you did not consider this case, please correct me if I am wrong. I'll give a possible flow below, this is similar to @benHeid 's advanced non-linear pipeline.

Suppose a forecasting pipeline has 3 steps: first it predicts few of the future-unknown `X` variables using other future-known `X` variables, then their union is used to predict `y`. The forecasters used in the two steps need not be same, and one may have native support for categorical features and other may not. In that case, it's perfectly reasonable for user to want to take benefit of native support when available and use encoding when not.

I am visualising that in this case we need to encode only for the forecaster that does not have native support, not for the other one. If there are more estimators in the pipeline where same variable(s) need to be encoded, reuse same trained encoder there too instead of training again and again. This does not have to be the only way, and there can be better methods, but as of now I am not sure if this can be handled if the encoding of categorical features happen outside of `sktime` pipeline by users directly.